app/src/main/java/com/example/woloxapp/ui/Home/tablayout/fragments/NewsFragment.kt

```diff
@@ -22,14 +26,88 @@ class NewsFragment : Fragment() {
     }
 
     override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        newsViewModel = ViewModelProvider(this)[NewsViewModel::class.java]
         with(binding) {
             recyclerNews.layoutManager = LinearLayoutManager(context)
-            recyclerNews.adapter = NewsAdapter(NewsProvider.newsList)
-            if (NewsProvider.newsList.isEmpty()) emptyNews.visibility =
-                View.VISIBLE
+            recyclerNews.adapter = NewsAdapter(mutableListOf())
+            newsViewModel.getData(false)
+            newsViewModel.newsResponse.observe(viewLifecycleOwner) {
+                it?.let {
+                    (recyclerNews.adapter as NewsAdapter).moreNews(it.page)
+                    (recyclerNews.adapter as NewsAdapter).notifyDataSetChanged()
+                }
+            }
+            recyclerNews.addOnScrollListener(
+                InnerScrollObserver() {
+                    newsViewModel.getData(false)
+                }
+            )
+            newsViewModel.newListOk.observe(viewLifecycleOwner) {
+                if (it != null) binding.progressBarNews.visibility =
+                    View.INVISIBLE
+                when (it) {
+                    NewsViewModel.ResponseStatus.NewListOk -> showEmptyTextView(
+                        false
+                    )
+                    NewsViewModel.ResponseStatus.NewListError,
+                    NewsViewModel.ResponseStatus.NewListFailure -> newListErrorMessages()
+                    null -> binding.progressBarNews.visibility = View.VISIBLE
+                }
+            }
             swipeRefresh.setOnRefreshListener {
-                swipeRefresh.isRefreshing = false
+                newsViewModel.getData(true)
+                newsViewModel.newListOk.observe(viewLifecycleOwner) {
+                    swipeRefresh.isRefreshing = it == null
+                }
+            }
+        }
+    }
+
+    inner class InnerScrollObserver(
+        private val getMoreData: () -> Unit
+    ) : RecyclerView.OnScrollListener() {
+        override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
+            super.onScrolled(recyclerView, dx, dy)
+            childCount = recyclerView.childCount
+            totalCount = recyclerView.layoutManager!!.itemCount
+            firstItem =
+                (recyclerView.layoutManager as LinearLayoutManager).findFirstVisibleItemPosition()
+            getNews =
+                !loadingNews && (totalCount - childCount) <= (firstItem + visiblePort)
```

Comment:
This is a great approach! But the ScrollObserver is doing way too much. No need to change this, but think of a way when its only work is to notify when it reached bottom. Then, from your view model you can handle something like
```
scrollNotified(){
   if(pageNotChanged) // we haven't finish the current request
      skip()
   else
      requestNewPage()
}
```